canonical/ubuntu-pro-client
# Prefer the 'apt_pkg' module over 'apt' (SC-1516)

## Why is this needed?
This PR solves all of our problems because we are trying to standardize the use of `python3-apt` through the codebase. After some discussions with @julian-klode in the past,  we have decided to stick to `apt_pkg` instead of `apt` when possible, so we would:
- improve performance
- rely on a single mechanism to mess with apt data
- guarantee that no hidden functionality in `apt` is breaking our isolation of configuration, given that we are responsible for the esm-apt instance
- open possibilities to implement more features (as reporting progress, for instance) using `apt_pkg` capabilities.

## Test Steps
This is just a medium-size refactor; no new features introduced, so CI should pass without major changes (although adjustments in unit tests were of course needed because we still test some implementation details and isolation is not 100%)

It is possible to verify performance gains, maybe where it is more impactful: run `pro security-status` on an updated unattached bionic (any release, really) machine both with and without this patch and compare the differences.

## Checklist

## Does this PR require extra reviews?
 - [x] Yes - it would be nice to have @julian-klode 's eyes on it before merging
